modules/gmc-agent/src/main/java/org/gridgain/action/controller/QueryActionsController.java

```diff
@@ -0,0 +1,185 @@
+        qryRegistry = new QueryHolderRegistry(ctx, Duration.ofMinutes(5));
+    }
+
+    /**
+     * Cancel query by query ID.
+     *
+     * @param qryId Query id.
+     */
+    public CompletableFuture<Void> cancel(String qryId) {
+        qryRegistry.cancelQuery(qryId);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * @param arg Argument.
+     * @return Next page with result.
+     */
+    public CompletableFuture<QueryResult> nextPage(NextPageQueryArgument arg) {
+        final CompletableFuture<QueryResult> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            try {
+                String qryId = arg.getQueryId();
+                if (F.isEmpty(qryId))
+                    throw new IllegalArgumentException("Failed to execute query due to empty query ID");
+
+                String cursorId = arg.getCursorId();
+                if (F.isEmpty(cursorId))
+                    throw new IllegalArgumentException("Failed to execute query due to empty cursor ID");
+
+                CursorHolder cursorHolder = qryRegistry.findCursor(qryId, cursorId);
+                QueryResult res = fetchResult(cursorHolder, arg.getPageSize());
+
+                if (!res.isHasMore())
+                    qryRegistry.closeQueryCursor(qryId, cursorId);
+
+                fut.complete(res);
+            }
+            catch (Throwable e) {
+                fut.completeExceptionally(e);
+            }
+        }, MANAGEMENT_POOL);
+
+        return fut;
+    }
+
+    /**
+     * @param arg Argument.
+     * @return List of query results.
+     */
+    public CompletableFuture<List<QueryResult>> executeSqlQuery(QueryArgument arg) {
+        final CompletableFuture<List<QueryResult>> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            String qryId = arg.getQueryId();
+            qryRegistry.cancelQuery(qryId);
+
+            QueryHolder qryHolder = qryRegistry.createQueryHolder(qryId);
+
+            try {
+                if (log.isDebugEnabled())
+                    log.debug("Execute query started with subject: " + ctx.security().securityContext().subject());
+
+                SqlFieldsQuery qry = prepareQuery(arg);
+                GridCacheContext cctx = !F.isEmpty(arg.getCacheName())
+                        ? ctx.cache().cache(arg.getCacheName()).context()
+                        : null;
+
+                List<QueryResult> results = new ArrayList<>();
+                for (FieldsQueryCursor<List<?>> cur : qryProc.querySqlFields(cctx, qry, null, true, false, qryHolder.getCancelHook())) {
+                    CursorHolder cursorHolder = new CursorHolder(cur);
+                    QueryResult res = fetchResult(cursorHolder, arg.getPageSize());
+
+                    if (res.isHasMore())
+                        res.setCursorId(qryRegistry.addCursor(qryId, cursorHolder));
+
+                    results.add(res.setQueryId(qryId));
```

Comment:
Why we need qryId in results if we pass it from frontend?

Reply:
Agree, remove.